twister: terminate_process: fix NoSuchProcess error

M1cha opened the pull request:
NOTE: Even though previous commits indicate, that this can only happen on MacOS, that's actually not true. It happens on Linux as well.

The constructor of `psutil.Process` can throw an exception as well, so we need to wrap the whole loop in another try, unfortunately.

nashif commented:
@gchwier  @PerMac  can you review please?

PerMac commented on scripts/pylib/pytest-twister-harness/src/twister_harness/device/utils.py:
```diff
@@ -41,11 +41,16 @@ def terminate_process(proc: subprocess.Popen) -> None:
-        except (ProcessLookupError, psutil.NoSuchProcess):
-            pass
+
+    try:
+        for child in psutil.Process(proc.pid).children(recursive=True):
+            try:
+                os.kill(child.pid, signal.SIGTERM)
+            except (ProcessLookupError, psutil.NoSuchProcess):
+                pass
+    except (ProcessLookupError, psutil.NoSuchProcess):
+        pass
```
The code will work but it's good to keep it relevant. One exception is targeting `os` and the other `psutil`. A relevant error should correspond to each of them (i.e. os won't raise psutill exception and vice versa) hence my suggestion:

```suggestion
    try:
        for child in psutil.Process(proc.pid).children(recursive=True):
            try:
                os.kill(child.pid, signal.SIGTERM)
            except (ProcessLookupError):
                pass
    except (psutil.NoSuchProcess):
        pass
```

Or in this case if we just want to pass over exception a `suppress` can be used:

```
    from contextlib import suppress
    
    with suppress(psutil.NoSuchProcess):            
        for child in psutil.Process(proc.pid).children(recursive=True):
            with suppress(ProcessLookupError):
                os.kill(child.pid, signal.SIGTERM)
```

M1cha replied:
that's not quite right. both `Process()` and `kill()` can raise both of these exceptions.

PerMac reviewed (changes requested):
The code will work but it is preferable to keep the code relavant. One try/except should address the os exception and the other the psutill one. Both cannot happen in the same place.

gchwier commented:
Please check https://github.com/zephyrproject-rtos/zephyr/pull/82043/commits/a8d8073bc3cfc988785bb7d408c0dd694694179d
where contextlib was also introduced. Shouldn't your changes also be applied in the `handlers.py` file where process termination is used by Twister?

M1cha committed:
twister: terminate_process: use contextlib.suppress

While it doesn't cause any issues, it's more consistent and makes future
commits which add handling for more exception types more readable.

Based on: 0df8240b491

Signed-off-by: Michael Zimmermann <sigmaepsilon92@gmail.com>

M1cha committed:
twister: terminate_process: fix NoSuchProcess error

NOTE: Even though previous commits indicate, that this can only happen on
MacOS, that's actually not true. It happens on Linux as well.

The constructor of `psutil.Process` can throw an exception as well, so we
need to wrap the whole loop in another try, unfortunately.

Signed-off-by: Michael Zimmermann <sigmaepsilon92@gmail.com>

M1cha commented:
> Note, that you can always re-open a closed pull request at any time.

That's not true, GitHub does not allow me to reopen this PR.

M1cha commented:
I need to comment to reopen

M1cha commented:
I need to comment to reopen
